### PR TITLE
layout: Convert all inline iteration to a new `foreach` function

### PR DESCRIPTION
**Pull request by mrobinson**
Instead of a tricky stack of enum iterators expose a `foreach()`
function on InlineFormattingContext, which takes a `FnMut`. This
prevents callers wanting to iterate from keeping a stack of iterators
and will potentially allow a future version of this function to avoid
borrowing the `ArcRefCell<...>` of inline boxes for every iteration
(presumably using something like `OwnedRef`).

Convert `inline_content_sizes` to use this new `foreach()` function and
move the `Computation` out of the function body to
`ContentSizesComputation`. This reduces the stack depth of inline size
computation, because `foreach()` is iterative and not recursive.

This is a preliminary change to removing the second round of text shaping
during layout, because shaping will use this new iterator.

---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: components/layout_2020/flow/inline.rs [hunk 1]
```diff
@@ -2389,3 +2246,161 @@ fn inline_container_needs_strut(
...
+                    .margin(self.containing_block_writing_mode)
+                    .percentages_relative_to(zero)
+                    .auto_is(Length::zero);
+
+                let pbm = &(&margin + &padding) + &border;
+                if inline_box.is_first_fragment {
+                    self.add_length(pbm.inline_start);
+                }
+                if inline_box.is_last_fragment {
+                    self.ending_inline_pbm_stack.push(pbm.inline_end);
```
Comment (Loirooriol) [inline components/layout_2020/flow/inline.rs:2293]: I don't really understand `ending_inline_pbm_stack`, why not just use `add_length` here?

Comment (mrobinson) [reply components/layout_2020/flow/inline.rs:2293]: This is so the length is added after the contents of the inline box, which might adjust the state of content sizes calculation. This is needed now that the calculation is not recursive any longer. The ending length is applied when the "EndInlineBox" item is processed.

Comment (mrobinson) [reply components/layout_2020/flow/inline.rs:2293]: This is so the length is added after the contents of the inline box, which might adjust the state of content sizes calculation. This is needed now that the calculation is not recursive any longer. The ending length is applied when the "EndInlineBox" item is processed.

**Review by Loirooriol — COMMENTED** [review]
Lots of tests are failing

**Review by Loirooriol — APPROVED** [review]

**Pull request merged**